core/iwasm/aot/debug/jit_debug.c

```diff
@@ -0,0 +1,224 @@
+
+typedef struct WASMJITEntryNode {
+    struct WASMJITEntryNode *next;
+    ;
+    JITCodeEntry *entry;
+} WASMJITEntryNode;
+
+static WASMJITDebugEngine *jit_debug_engine;
+
+static WASMJITDebugEngine *
+wasm_jit_debug_engin_create()
+{
+    WASMJITDebugEngine *engine;
+    if (!(engine = wasm_runtime_malloc(sizeof(WASMJITDebugEngine)))) {
+        LOG_ERROR("WASM Debug Engine error: failed to allocate memory");
+        return NULL;
+    }
+    memset(engine, 0, sizeof(WASMJITDebugEngine));
+    return engine;
+}
+
+void
+initial_jit_debug_engine()
+{
+    if (!jit_debug_engine)
+        jit_debug_engine = wasm_jit_debug_engin_create();
+    os_mutex_init(&jit_debug_engine->jit_entry_mutex);
+    bh_list_init(&jit_debug_engine->jit_entry_list);
+}
+
+static JITCodeEntry *
+CreateJITCodeEntryInternal(const uint8_t *symfile_addr, uint64_t symfile_size)
+{
+    JITCodeEntry *entry;
+
+    if (!jit_debug_engine)
+        return NULL;
+
+    os_mutex_lock(&jit_debug_engine->jit_entry_mutex);
+    if (!(entry = wasm_runtime_malloc(sizeof(JITCodeEntry)))) {
+        LOG_ERROR("WASM JIT Debug Engine error: failed to allocate memory");
+        os_mutex_unlock(&jit_debug_engine->jit_entry_mutex);
+        return NULL;
+    }
+    entry->symfile_addr_ = symfile_addr;
+    entry->symfile_size_ = symfile_size;
+    entry->prev_ = NULL;
+
+    entry->next_ = __jit_debug_descriptor.first_entry_;
+    if (entry->next_ != NULL) {
+        entry->next_->prev_ = entry;
+    }
+    __jit_debug_descriptor.first_entry_ = entry;
+    __jit_debug_descriptor.relevant_entry_ = entry;
+
+    __jit_debug_descriptor.action_flag_ = JIT_REGISTER_FN;
+
+    (*__jit_debug_register_code_ptr)();
+    os_mutex_unlock(&jit_debug_engine->jit_entry_mutex);
+    return entry;
+}
+
+static void
+DeleteJITCodeEntryInternal(JITCodeEntry *entry)
+{
+    if (!jit_debug_engine)
+        return;
+    os_mutex_lock(&jit_debug_engine->jit_entry_mutex);
+    if(entry->prev_ != NULL) { entry->prev_->next_ = entry->next_; }
+    else { __jit_debug_descriptor.first_entry_ = entry->next_; }
+    if (entry->next_ != NULL) {
+        entry->next_->prev_ = entry->prev_;
+    }
+    __jit_debug_descriptor.relevant_entry_ = entry;
+    __jit_debug_descriptor.action_flag_ = JIT_UNREGISTER_FN;
+    (*__jit_debug_register_code_ptr)();
+    wasm_runtime_free(entry);
+    os_mutex_unlock(&jit_debug_engine->jit_entry_mutex);
+}
+
+static void
+test_save_symfile(const uint8_t *symfile_addr, uint64_t symfile_size)
+{
+    int fd;
+    fd = open("test.o", O_RDWR | O_SYNC | O_CREAT, 0644);
+    write(fd, symfile_addr, symfile_size);
+    fsync(fd);
+}
+void
+create_jit_code_entry(const uint8_t *symfile_addr, uint64_t symfile_size)
+{
+    JITCodeEntry *entry;
+    WASMJITEntryNode *node;
+
+    if (!jit_debug_engine)
+        return;
+
+    if (!(node = wasm_runtime_malloc(sizeof(WASMJITEntryNode)))) {
+        LOG_ERROR("WASM JIT Debug Engine error: failed to allocate memory");
+        return;
```

Comment:
Should return false to indicate the caller that create failed so that caller can handle the error.